internal/database/badger/writebatch.go

```diff
@@ -0,0 +1,40 @@
+package badger
+
+import "github.com/dgraph-io/badger/v3"
+
+// writeBatch uses the badger write batch and prefixes
+// all keys with a certain given prefix.
+type writeBatch struct {
+	prefix           []byte
+	badgerWriteBatch *badger.WriteBatch
+}
+
+func newWriteBatch(prefix []byte, badgerWriteBatch *badger.WriteBatch) *writeBatch {
+	return &writeBatch{
+		prefix:           prefix,
+		badgerWriteBatch: badgerWriteBatch,
+	}
+}
+
+// Set sets a value at the given key prefixed with the given prefix.
+func (wb *writeBatch) Set(key, value []byte) (err error) {
+	key = append(wb.prefix, key...)
+	return wb.badgerWriteBatch.Set(key, value)
```

Comment:
here we use the badger db writer batch instead of writing our own (chaindb one is faulty on top of that)